### PR TITLE
add block-level locking (WP-530)

### PR DESCRIPTION
**Pull request by vsolovei-smartling**

### CODE REVIEW

**Comment by vsolovei-smartling** [comment]
I intend to add/change tests, but hopefully no more major changes

File: inc/Smartling/WP/Controller/TranslationLockController.php [hunk 1]
```diff
@@ -153,26 +155,19 @@ public function popupIFrame()
...
-            $submission = $this->getSubmissionFromQuery();
-            if (false !== $submission) {
-                $submission->setLockedFields(serialize($_lockedFields));
-                $submission->setIsLocked($_pageLock);
-                $this->getManager()->storeEntity($submission);
-            }
-
+        $submission = $this->getSubmissionFromQuery();
+        if (false !== $submission) {
+            $submission->setLockedFields(serialize(array_key_exists('lockField', $_POST) && is_array($_POST['lockField']) ? array_keys($_POST['lockField']) : []));
```
Comment (dimitrystd) [inline inc/Smartling/WP/Controller/TranslationLockController.php:168]: I see a lot of inlinings in this PR. For this particular example the old code was more clear than this one.

**Review by dimitrystd — COMMENTED** [review]
It seems this release has no changes in XML, which is good. But it has limitations by design (blocks reordering; more or less blocks). As for me, it's a major flaw for real-world use cases

File: inc/Smartling/DbAl/WordpressContentEntities/EntityAbstract.php [hunk 2]
```diff
@@ -158,6 +158,7 @@ protected function getFieldNameByMethodName($method)
      */
     public function __call($method, array $params)
     {
+        $this->getLogger()->debug("Magic call to $method");
```
[the remark this answers is not in the document]
Comment (vsolovei-smartling) [reply inc/Smartling/DbAl/WordpressContentEntities/EntityAbstract.php:161]: As a long term goal, I would like to get rid of this magic method, but to do so, we need to see if it is used.

File: inc/Smartling/WP/Controller/TranslationLockController.php (hunk 1 above)
Comment (vsolovei-smartling) [reply inc/Smartling/WP/Controller/TranslationLockController.php:168]: Reverted

**Review by PavelLoparev — COMMENTED** [review]
1. This `applyXml` method is already too big and we extend it even more. Seems like this "lock Guttenberg module/block" functionality can be implemented as action handler. For instance, there is a `EVENT_SMARTLING_AFTER_DESERIALIZE_CONTENT` being thrown from `applyXml` method right before your custom code. Maybe it would be better to implement the whole class as an action handler?
2. Tests are missing.

File: inc/Smartling/Helpers/XmlHelper.php [hunk 3]
```diff
@@ -113,10 +113,9 @@ public static function xmlEncode(array $source, SubmissionEntity $submission, ar
         static::logMessage(vsprintf('Started creating XML for fields: %s', [base64_encode(var_export($source, true))]));
         $xml = self::setTranslationComments(self::initXml());
         $settings = self::getFieldProcessingParams();
-        $keySettings = &$settings['key'];
```
[the remark this answers is not in the document]
Comment (vsolovei-smartling) [reply inc/Smartling/Helpers/XmlHelper.php:116]: The settings variable is not used in this metod, nor changed in the only call to rowToXMLNode

File: inc/Smartling/Helpers/PostContentHelper.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+     * @param string $translated
+     * @param array $lockedFields
+     * @return string
+     */
+    public function applyTranslation($original, $translated, array $lockedFields)
+    {
+        $result = $this->blockHelper->normalizeCoreBlocks($translated);
+        $originalBlocks = $this->blockHelper->addPostContentBlocks(['post_content' => $original]);
+        $translatedBlocks = $this->blockHelper->addPostContentBlocks(['post_content' => $translated]);
+        unset ($originalBlocks['post_content'], $translatedBlocks['post_content']);
```
[the remark this answers is not in the document]
Comment (vsolovei-smartling) [reply inc/Smartling/Helpers/PostContentHelper.php:26]: This method adds blocks to post content, but we don't need the post content itself here

File: inc/Smartling/Submissions/SubmissionEntity.php [hunk 5]
```diff
@@ -356,9 +356,7 @@ public function setStatus($status)
 
     public function hasLocks()
     {
-        $fields = maybe_unserialize($this->getLockedFields());
-
-        return 1 === $this->getIsLocked() || (is_array($fields) && 0 < count($fields));
+        return 1 === $this->getIsLocked() || 0 < count($this->getLockedFields());
```
[the remark this answers is not in the document]
Comment (vsolovei-smartling) [reply inc/Smartling/Submissions/SubmissionEntity.php:359]: The call is moved to getLockedFields

File: inc/Smartling/Base/SmartlingCoreUploadTrait.php [hunk 6]
```diff
@@ -191,30 +178,24 @@ private function readLockedTranslationFieldsBySubmission(SubmissionEntity $submi
...
+        $targetContent = $this->getContentHelper()->readTargetContent($submission)->toArray();
         $targetMeta = $this->getContentHelper()->readTargetMetadata($submission);
 
         $this->getLogger()->debug(vsprintf('Got target metadata: %s.', [var_export($targetMeta, true)]));
 
-        foreach ($lockedFields as $lockedFieldName) {
-
-            if (preg_match('/^meta\//ius', $lockedFieldName)) {
-                $_fieldName = preg_replace('/^meta\//ius', '', $lockedFieldName);
+        foreach ($submission->getLockedFields() as $lockedFieldName) {
```
[the remark this answers is not in the document]
Comment (vsolovei-smartling) [reply inc/Smartling/Base/SmartlingCoreUploadTrait.php:190]: The maybe_unserialize call now happens inside $submission->getLockedFields()

**Comment by vsolovei-smartling** [comment]
> 1. This `applyXml` method is already too big and we extend it even more. Seems like this "lock Guttenberg module/block" functionality can be implemented as action handler. For instance, there is a `EVENT_SMARTLING_AFTER_DESERIALIZE_CONTENT` being thrown from `applyXml` method right before your custom code. Maybe it would be better to implement the whole class as an action handler? For now it looks like we don't use own api.
The api is meant for third-party extension of the connector plugin, isn't it?

File: inc/Smartling/Submissions/SubmissionEntity.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (PavelLoparev) [reply inc/Smartling/Submissions/SubmissionEntity.php:359]: Didn't catch this. Ok

**Comment by PavelLoparev** [comment]
> The api is meant for third-party extension of the connector plugin, isn't it?
Yes. And I don't see any reasons not to use it by connector instead of adding more conditions to the main flow.

@dimitrystd what do you think?

**Review by PavelLoparev — APPROVED** [review]

**Comment by PavelLoparev** [comment]
I think the main question is

>But it has limitations by design (blocks reordering; more or less blocks). As for me, it's a major flaw for real-world use cases

Will blocks work after reordering/adding/removing blocks?

**Pull request merged**